src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/DocumentPresentation/TextDocumentUriPresentationEndpoint.cs

```diff
@@ -0,0 +1,225 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.AspNetCore.Razor.LanguageServer.Common.Extensions;
+using Microsoft.AspNetCore.Razor.LanguageServer.Extensions;
+using Microsoft.AspNetCore.Razor.LanguageServer.ProjectSystem;
+using Microsoft.AspNetCore.Razor.LanguageServer.Protocol;
+using Microsoft.CodeAnalysis.Razor;
+using Microsoft.Extensions.Logging;
+using OmniSharp.Extensions.LanguageServer.Protocol.Models;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer.DocumentPresentation
+{
+    internal class TextDocumentUriPresentationEndpoint : ITextDocumentUriPresentationHandler
+    {
+        private readonly ProjectSnapshotManagerDispatcher _projectSnapshotManagerDispatcher;
+        private readonly DocumentResolver _documentResolver;
+        private readonly RazorDocumentMappingService _razorDocumentMappingService;
+        private readonly RazorComponentSearchEngine _razorComponentSearchEngine;
+        private readonly ILogger _logger;
+
+        public TextDocumentUriPresentationEndpoint(
+            ProjectSnapshotManagerDispatcher projectSnapshotManagerDispatcher,
+            DocumentResolver documentResolver,
+            RazorDocumentMappingService razorDocumentMappingService,
+            RazorComponentSearchEngine razorComponentSearchEngine,
+            ILoggerFactory loggerFactory)
+        {
+            if (projectSnapshotManagerDispatcher is null)
+            {
+                throw new ArgumentNullException(nameof(projectSnapshotManagerDispatcher));
+            }
+
+            if (documentResolver is null)
+            {
+                throw new ArgumentNullException(nameof(documentResolver));
+            }
+
+            if (razorDocumentMappingService is null)
+            {
+                throw new ArgumentNullException(nameof(razorDocumentMappingService));
+            }
+
+            if (razorComponentSearchEngine is null)
+            {
+                throw new ArgumentNullException(nameof(razorComponentSearchEngine));
+            }
+
+            if (loggerFactory is null)
+            {
+                throw new ArgumentNullException(nameof(loggerFactory));
+            }
+
+            _projectSnapshotManagerDispatcher = projectSnapshotManagerDispatcher;
+            _documentResolver = documentResolver;
+            _razorDocumentMappingService = razorDocumentMappingService;
+            _razorComponentSearchEngine = razorComponentSearchEngine;
+            _logger = loggerFactory.CreateLogger<TextDocumentUriPresentationEndpoint>();
+        }
+
+        public RegistrationExtensionResult GetRegistration()
```

Comment:
You'll probably have merge conflicts with the bits I just merged this morning. Should be an easy change though